### PR TITLE
Warn about unused preprocessor_deps

### LINKED ISSUE #2642: (preprocessor_deps) are ignored when there is no preprocessing
Hi,

I was trying to use `(preprocessor_deps)` as a way to pull files (required at runtime) into `_build/`, and noticed the following behaviour.

When there is no `(preprocess)` field, `(preprocessor_deps)` is completely ignored.

For example, the following `dune` file does not trigger any error or warning (but uncommenting the `(preprocess)` field triggers an error if the file does not exist of course).

```scheme
(executable
 (name e)
 ;(preprocess (pps ppx_deriving.std))
 (preprocessor_deps does-not-exist.txt))
```

I think that we could detect the problem a bit more eagerly, for example by issuing an error or a warning when `(preprocess_deps)` is set but `(preprocess)` is not.

What do you think?

### PR DESCRIPTION
**Pull request by Julow**
Fix https://github.com/ocaml/dune/issues/2642

Show a warning when a `preprocessor_deps` field is provided but `preprocess` is not:

```
File "dune", line 5, characters 1-39:
5 |  (preprocessor_deps does-not-exist.txt))
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Warning: This preprocessor_deps field will be ignored because no preprocessor
is configured.
```

How to make sure this is turned into an error with lang dune >= 2.0 ?

### CODE REVIEW

**Comment by nojb** [comment]
> How to make sure this is turned into an error with lang dune >= 2.0 ?

`User_warning.emit` has a `~is_error` optional argument, and you can get the current version from `Stanza.syntax`.

**Comment by Julow** [comment]
Thanks. I implemented it using the current project's `dune_version` instead. What do you think?

File: src/dune/preprocessing.ml [hunk 1]
```diff
@@ -566,20 +566,36 @@ type t = (Module.t -> lint:bool -> Module.t) Per_module.t
...
     ~lib_name ~scope =
+  let preprocess =
+    Per_module.map preprocess ~f:(fun pp ->
+        Dune_file.Preprocess.remove_future_syntax ~for_:Compiler pp
+          (Super_context.context sctx).version)
+  in
   let preprocessor_deps =
-    Build.memoize "preprocessor deps" preprocessor_deps
+    let loc, deps = preprocessor_deps in
+    ( if (not (is_preprocessing preprocess)) && not (List.is_empty deps) then
```
Comment (nojb) [inline src/dune/preprocessing.ml]: parentheses around the first `not` are not needed.

**Review by nojb — APPROVED** [review]
LGTM

File: src/dune/preprocessing.ml (hunk 1 above)
Comment (Julow) [reply src/dune/preprocessing.ml]: Fixed.

**Comment by nojb** [comment]
Thanks!

**Comment by nojb** [comment]
I think we forgot a CHANGES entry. Am a bit busy at the moment, do you mind making a PR?

**Comment by diml** [comment]
I updated the changelog in master and pushed a couple of tweaks: morally we want to use `Stanza.syntax` because different part of the grammar might be attached to different syntaxes. I also moved the check at parsing time, since this is where we usually check syntactic properties. I just disallowed to have a `preprocessor_deps` field with `future_syntax` given that `future_syntax` cannot have runtime dependencies.

**Pull request merged**